<id>hlp.svc-align-tophat.html v2.2.1</id>
<title>TopHat Alignment Tool</title>
<definition>TopHat algorithm description</definition>
<description>
TopHat is a program that aligns RNA-Seq reads 
to a genome in order to identify exon-exon 
splice junctions. It is built on the ultrafast 
short read mapping program Bowtie. TopHat runs 
on Linux and OS X, but HIVE-integrated TopHat 
alignment is optimized to work in a highly 
parallelized fashion within the HIVE environment.
To access the TopHat help document, 
<a href="http://ccb.jhu.edu/software/tophat/manual.shtml">click here</a>.
<br/>
<br/>
All of the Align-X (externally adapted) aligners that 
have been integrated into HIVE use the underlying 
HIVE-Hexagon interface and therefore have the same 
set of minimum required inputs that can be viewed 
<a href="hlp.svc-align-hexagon.html">here</a> or 
below. However, TopHat has a few additional, optional 
parameters that are described below, along with 
standard HIVE-Hexagon input parameters:
<br/>
<dl>
    <dt><span class='pn'>Name</span></dt>
        <dd>
            Here you can designate a name for this specific 
            alignment process for simple search and retrieval from 
            your user home directory.
        </dd>
<br/>
    <dt><span class='pn'>Alignment Algorithm</span></dt>
        <dd>
            Here you may choose the algorithm used for alignment, the 
            default selection is HIVE-hexagon. Other selections include 
            NCBI BLAST, Bowtie, BWA and the UCSC aligner BLAT. Select the 
            desired algorithm by clicking on the associated title or 
            description field.
        </dd>
<br/>
    <dt><span class='pn'>Sequence Read</span></dt>
        <dd>
            Select the query data for your alignment by clicking the arrow. 
            A new window will pop up displaying all reads available to you 
            through your user home directory. Select one or multiple reads 
            by clicking on the corresponding row. To close the window click 
            the <img src="img/use-delete.gif" width=15> icon in the top 
            right corner. Your selected sequence(s) should now be visible 
            in the Sequence Read field. Sequence reads can also be 
            pre-populated if selected directly from your user home 
            directory prior to clicking the HIVE-hexagon shortcut 
            <img src="img/hiveicon.jpg" width=15> icon.
        </dd>
<br>
    <dt><span class='pn'>Reference Genome</span></dt>
        <dd>
            Select the genome or set to be used as the reference for your 
            alignment by clicking the arrow in the corresponding field. A new 
            window will open displaying all genome objects available to you. 
            Select one or multiple references by clicking the containing row. 
            To close the window click the <img src="img/use-delete.gif" width=15> 
            icon in the top right corner. Your selected reference object(s) 
            should appear in the Reference Genome field. Like reads, references 
            can be pre-populated if selected from your user home directory and 
            accessing HIVE-hexagon through the shortcut by clicking the 
            <img src="img/hiveicon.jpg" width=15> icon.
        </dd>
<br/>
    <dt><span class='pn'>GTF file</span></dt>
        <dd>
            Supply TopHat with a set of gene model annotations 
            and/or known transcripts, as a GTF 2.2 or GFF3 
            formatted file. If this option is provided, TopHat 
            will first extract the transcript sequences and use 
            Bowtie to align reads to this virtual transcriptome 
            first. Only the reads that do not fully map to the 
            transcriptome will then be mapped on the genome. The 
            reads that did map on the transcriptome will be 
            converted to genomic mappings (spliced as needed) and 
            merged with the novel mappings and junctions in the 
            final tophat output.
            <br/><br/>
            Please note that the values in the first column of the 
            provided GTF/GFF file (column which indicates the 
            chromosome or contig on which the feature is located), 
            must match the name of the reference sequence in the 
            Bowtie index you are using with TopHat. Before using a 
            known annotation file with this option please make sure 
            that the 1st column in the annotation file uses the 
            exact same chromosome/contig names (case sensitive) as 
            the names of the reference sequences in the Bowtie 
            index you are using.
        </dd>
<br/>
    <dt><span class='pn'>Minimum Match Length</span></dt>
        <dd>
            The alignments produced may vary in length. This option allows the 
            user to choose a preferred length of alignment by discarding all 
            alignments shorter in length than the specified value.
        </dd>
<br/>        
    <dt><span class='pn'>Matches to Keep</span></dt>
        <dd>
            Reported alignments can be filtered by choosing one of five options 
            related to the associated score of the alignments:
            <ul>
                <li><i>First Match</i> reports the first hit found by Hexagon 
                regardless of optimization or scoring</li>
                <li><i>Best first match</i> will take a small sampling of initial 
                matches and report the highest scoring hit of those found</li>
                <li><i>All matches within acceptable thresholds</i> will report any 
                alignment found which meets the criteria specified at job request</li>
                <li><i>All equally best alternative matches</i> will find all possible 
                optimal alignments for a given short read but report only those found 
                with the highest score</li>
                <li><i>Random vote between equally best alternative matches</i>, the 
                default selection, will again find all possible optimal alignments 
                for a given read but will randomly choose and report a single hit 
                from the pool of highest scoring alignments</li>
            </ul>
        </dd>
<br/>
    <dt><span class='pn'>Mismatches</span></dt> 
        <dd>
            A mismatch in an alignment occurs when the query and reference align at most points 
            in a given region but have different nucleotides at a single position in the same 
            region. Potential alignments are not discarded due solely to mismatches, but the 
            cumulative alignment score reflects a penalty associated with mismatches. The Mismatch 
            Percentage is equal to the total number of mismatches divided by the total number of 
            positions (or Mismatch % = total # mismatches/total # positions). This filter removes 
            alignments with a mismatch % greater than the specified value.
            <ul>
                <li><i>Percent Allowed</i> allows the user to designate the percentage 
                of mismatches allowed over which alignments will not be reported.</li>
                <li><i>Computed on</i> allows the user to designate whether to compute 
                the mismatch percent based on the length of the query, the entire 
                alignment, or the designated minimum match length.</li>
            </ul>
        </dd>
<br/>
    <dt><span class='pn'>Batch Service</span></dt>
        <dd>
            This parameter allows submission of alignment jobs as 
            individual alignments or as a batch of several alignments. 
            Selecting batch mode from the dropdown menu will reload 
            the HIVE-hexagon tool page for use in batch mode. For 
            more details, 
            <a href="hlp.obj.batch-params.html">please click here to 
            see the batch parameters advanced help topic</a>.
        </dd>
</dl>
</ul>

</description>
